### ISSUE TITLE
Broken Tree Shaking - Icons

### ISSUE THREAD

**Issue by montezume** [commercetools/ui-kit#307]
#### Background

Most of our bundle size is currently consumed with base64 encoded icon svgs.

#### Current situation

If a user does the following import in their application
```js
import { AddBoldIcon } from '@commercetools-frontend/ui-kit'
```

their bundle is bloated with all of the Icons contained in ui-kit. 

#### Definition of done

Users are able to import one icon, without having their bundle bloated with all of our icons.

**Comment by emmenko** [comment]
Hmm, are you sure that tree-shaking is not working? Each icon is exported individually...could you clarify this problem a bit more?

**Comment by montezume** [comment]
@emmenko yeah, I spent quite a bit of time with it. I would imagine it has something to do with the generated Icons file. 

When webpack sees

```
const CubeIcon = withThemeProp(withSizeProp(OrigCubeIcon));
```

it has no way of knowing that withThemeProp and withSizeProp have no side effects.

**Comment by dferber90** [comment]
I believe that if we created one JS file per component instead of one JS file for all components the treeshaking would work. Each of the files could still import&use `withThemeProp` and `withSizeProp` from some utils file.

**Comment by emmenko** [comment]
> I believe that if we created one JS file per component instead of one JS file for all components the treeshaking would work.

Yeah, that might be an option 👌

**Comment by montezume** [comment]
@emmenko @dferber90 I tried this manually locally and the result is still the same 😕

**Comment by dferber90** [comment]
@emmenko Is this something you want to work on?

**Comment by montezume** [comment]
@emmenko 

index.js
```
export { AddBoldIcon } from './icons/components/add-bold-icon'
export { AddIcon } from './icons/components/add-icon'
```

ui-kit.esm.js
```js
// other stuff
function _extends$1() { _extends$1 = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends$1.apply(this, arguments); }

var _ref =
/*#__PURE__*/
React.createElement("g", {
  id: "add_react_svg__Icons",
  stroke: "none",
  strokeWidth: 1,
  fillRule: "evenodd"
}, React.createElement("g", {
  id: "add_react_svg__MC-icon-set",
  transform: "translate(-168 -120)",
  fill: "#000"
}, React.createElement("g", {
  id: "add_react_svg__CRUD",
  transform: "translate(24 120)"
}, React.createElement("g", {
  id: "add_react_svg__Add",
  transform: "translate(144)"
}, React.createElement("path", {
  d: "M12.5 11.5v-8a.5.5 0 1 0-1 0v8h-8a.5.5 0 1 0 0 1h8v8a.5.5 0 1 0 1 0v-8h8a.5.5 0 1 0 0-1h-8z",
  id: "add_react_svg__shape"
})))));

var SvgAddReact = function SvgAddReact(props) {
  return React.createElement("svg", _extends$1({
    width: 24,
    height: 24,
    viewBox: "0 0 24 24"
  }, props), _ref);
};

var AddIcon = withThemeProp(withSizeProp(SvgAddReact));
AddIcon.displayName = 'AddIcon';

function _extends$2() { _extends$2 = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends$2.apply(this, arguments); }

var _ref$1 =
/*#__PURE__*/
React.createElement("g", {
  id: "add-bold_react_svg__Icons",
  stroke: "none",
  strokeWidth: 1,
  fillRule: "evenodd"
}, React.createElement("g", {
  id: "add-bold_react_svg__MC-icon-set",
  transform: "translate(-240 -120)",
  fillRule: "nonzero",
  fill: "#000"
}, React.createElement("g", {
  id: "add-bold_react_svg__CRUD",
  transform: "translate(24 120)"
}, React.createElement("g", {
  id: "add-bold_react_svg__Add-Bold",
  transform: "translate(216)"
}, React.createElement("path", {
  d: "M13.844 10.334v-3.51c0-.97-.784-1.755-1.75-1.755-.967 0-1.75.785-1.75 1.755v3.51h-3.5c-.967 0-1.75.785-1.75 1.754 0 .97.783 1.755 1.75 1.755h3.5v3.51c0 .97.783 1.755 1.75 1.755.966 0 1.75-.786 1.75-1.755v-3.51h3.5c.966 0 1.75-.785 1.75-1.755 0-.969-.784-1.754-1.75-1.754h-3.5z",
  id: "add-bold_react_svg__shape"
})))));

var SvgAddBoldReact = function SvgAddBoldReact(props) {
  return React.createElement("svg", _extends$2({
    width: 24,
    height: 24,
    viewBox: "0 0 24 24"
  }, props), _ref$1);
};

var AddBoldIcon = withThemeProp(withSizeProp(SvgAddBoldReact));
AddBoldIcon.displayName = 'AddBoldIcon';

export { AddBoldIcon, AddIcon };

```

my app code
```js
import { AddBoldIcon } from "@commercetools-frontend/ui-kit";
console.log("AddBoldIcon", AddBoldIcon);
```

my built app code (webpack treeshaking requires minification)
```js
!function(n){var r={};function o(e){if(r[e])return r[e].exports;var t=r[e]={i:e,l:!1,exports:{}};return n[e].call(t.exports,t,t.exports,o),t.l=!0,t.exports}o.m=n,o.c=r,o.d=function(e,t,n){o.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:n})},o.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},o.t=function(t,e){if(1&e&&(t=o(t)),8&e)return t;if(4&e&&"object"==typeof t&&t&&t.__esModule)return t;var n=Object.create(null);if(o.r(n),Object.defineProperty(n,"default",{enumerable:!0,value:t}),2&e&&"string"!=typeof t)for(var r in t)o.d(n,r,function(e){return t[e]}.bind(null,r));return n},o.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return o.d(t,"a",t),t},o.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},o.p="",o(o.s=3)}([function(e,t){e.exports=React},function(e,t){e.exports=recompose},function(e,t){e.exports=classnames},function(e,t,n){"use strict";n.r(t);var r=n(2),o=n.n(r),a=n(1),c="Invariant failed";var i=function(e,t){if(!e)throw new Error(c)},l=n(0),s=n.n(l);function d(o){for(var e=1;e<arguments.length;e++){var a=null!=arguments[e]?arguments[e]:{},t=Object.keys(a);"function"==typeof Object.getOwnPropertySymbols&&(t=t.concat(Object.getOwnPropertySymbols(a).filter(function(e){return Object.getOwnPropertyDescriptor(a,e).enumerable}))),t.forEach(function(e){var t,n,r;t=o,r=a[n=e],n in t?Object.defineProperty(t,n,{value:r,enumerable:!0,configurable:!0,writable:!0}):t[n]=r})}return o}function _(e,t){if(null==e)return{};var n,r,o=function(e,t){if(null==e)return{};var n,r,o={},a=Object.keys(e);for(r=0;r<a.length;r++)n=a[r],0<=t.indexOf(n)||(o[n]=e[n]);return o}(e,t);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);for(r=0;r<a.length;r++)n=a[r],0<=t.indexOf(n)||Object.prototype.propertyIsEnumerable.call(e,n)&&(o[n]=e[n])}return o}var f={"theme-black":"icons-mod_theme-black__bZPye","theme-grey":"icons-mod_theme-grey__cZeDf","theme-white":"icons-mod_theme-white__1aGsC","theme-blue":"icons-mod_theme-blue__3B48t","theme-green":"icons-mod_theme-green__2_a6c","theme-green-light":"icons-mod_theme-green-light__ssJxQ","theme-orange":"icons-mod_theme-orange__2N9_B","theme-red":"icons-mod_theme-red__3i-RE",scale:"icons-mod_scale__2O4kX"};!function(e,t){void 0===t&&(t={});var n=t.insertAt;if(e&&"undefined"!=typeof document){var r=document.head||document.getElementsByTagName("head")[0],o=document.createElement("style");o.type="text/css","top"===n&&r.firstChild?r.insertBefore(o,r.firstChild):r.appendChild(o),o.styleSheet?o.styleSheet.cssText=e:o.appendChild(document.createTextNode(e))}}(".icons-mod_theme-black__bZPye *{\n  fill:#1a1a1a;\n}\n\n.icons-mod_theme-grey__cZeDf *{\n  fill:hsl(0, 0%, 60%);\n}\n\n.icons-mod_theme-white__1aGsC *{\n  fill:#fff;\n}\n\n.icons-mod_theme-blue__3B48t *{\n  fill:#078cdf;\n}\n\n.icons-mod_theme-green__2_a6c *{\n  fill:#00b39e;\n}\n\n.icons-mod_theme-green-light__ssJxQ *{\n  fill:hsl(172.9608938547486, 100%, 40%);\n}\n\n.icons-mod_theme-orange__2N9_B *{\n  fill:#f16d0e;\n}\n\n.icons-mod_theme-red__3i-RE *{\n  fill:#e60050;\n}\n\nsvg.icons-mod_scale__2O4kX:not(:root){\n  width:100%;\n  height:auto;\n}\n");var u=12,m=16,h=24,g=function(e){switch(e){case"small":return u;case"medium":return m;case"big":default:return h}},p=Object(a.mapProps)(function(e){var t=e.size,n=_(e,["size"]);return d({},n,"scale"===t?{className:o()(n.className,f.scale)}:{},{width:g(t),height:g(t)})}),v=Object(a.mapProps)(function(e){var t=e.theme;return d({},_(e,["theme"]),{className:function(e){if(e){var t=f["theme-".concat(e)];return i(t,"ui-kit/Icon: the specified theme '".concat(e,"' is not supported.")),t}}(t)})});function b(){return(b=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var n=arguments[t];for(var r in n)Object.prototype.hasOwnProperty.call(n,r)&&(e[r]=n[r])}return e}).apply(this,arguments)}var y=s.a.createElement("g",{id:"add_react_svg__Icons",stroke:"none",strokeWidth:1,fillRule:"evenodd"},s.a.createElement("g",{id:"add_react_svg__MC-icon-set",transform:"translate(-168 -120)",fill:"#000"},s.a.createElement("g",{id:"add_react_svg__CRUD",transform:"translate(24 120)"},s.a.createElement("g",{id:"add_react_svg__Add",transform:"translate(144)"},s.a.createElement("path",{d:"M12.5 11.5v-8a.5.5 0 1 0-1 0v8h-8a.5.5 0 1 0 0 1h8v8a.5.5 0 1 0 1 0v-8h8a.5.5 0 1 0 0-1h-8z",id:"add_react_svg__shape"})))));function O(){return(O=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var n=arguments[t];for(var r in n)Object.prototype.hasOwnProperty.call(n,r)&&(e[r]=n[r])}return e}).apply(this,arguments)}v(p(function(e){return s.a.createElement("svg",b({width:24,height:24,viewBox:"0 0 24 24"},e),y)})).displayName="AddIcon";var j=s.a.createElement("g",{id:"add-bold_react_svg__Icons",stroke:"none",strokeWidth:1,fillRule:"evenodd"},s.a.createElement("g",{id:"add-bold_react_svg__MC-icon-set",transform:"translate(-240 -120)",fillRule:"nonzero",fill:"#000"},s.a.createElement("g",{id:"add-bold_react_svg__CRUD",transform:"translate(24 120)"},s.a.createElement("g",{id:"add-bold_react_svg__Add-Bold",transform:"translate(216)"},s.a.createElement("path",{d:"M13.844 10.334v-3.51c0-.97-.784-1.755-1.75-1.755-.967 0-1.75.785-1.75 1.755v3.51h-3.5c-.967 0-1.75.785-1.75 1.754 0 .97.783 1.755 1.75 1.755h3.5v3.51c0 .97.783 1.755 1.75 1.755.966 0 1.75-.786 1.75-1.755v-3.51h3.5c.966 0 1.75-.785 1.75-1.755 0-.969-.784-1.754-1.75-1.754h-3.5z",id:"add-bold_react_svg__shape"}))))),w=v(p(function(e){return s.a.createElement("svg",O({width:24,height:24,viewBox:"0 0 24 24"},e),j)}));w.displayName="AddBoldIcon",console.log("AddBoldIcon",w)}]);
```

**Comment by emmenko** [comment]
Ah good, now it's much easier to "follow the problem".

With that said, I have no idea why that is...maybe we should understand first if the problem lies on rollup on webpack?

**Comment by emmenko** [comment]
I did some digging and experimented a bit with some dummy components `<Foo>` and `<Bar>`, with `<Bar>` importing a couple of icons. Then I tested using a fresh CRA the bundled app, importing only `<Foo>` and the resulting bundle contained the icons of `<Bar>`, which were supposed to be removed.
Same thing when bundling an app with Parcel.

- Webpack (CRA)
![screenshot 2019-01-28 at 10 06 43](https://user-images.githubusercontent.com/1110551/51825276-b53be500-22e4-11e9-9da4-8b4a7aac5433.png)

- Parcel
![screenshot 2019-01-28 at 10 10 03](https://user-images.githubusercontent.com/1110551/51825457-195ea900-22e5-11e9-87f0-b3baa3390721.png)

---

Probably there are other dependencies that will cause the same problems (e.g. I think prop-types also does it) or things like ES6 classes, static properties, etc.
It seems to me that it's still early days for having tree-shaking properly working.
I looked also at other libraries, issues, etc and there still seem to be a lot of problems or confusion on how to do it. Bundlers like Webpack also seem to have a "fragile" implementation, meaning that it will bail out as soon as it can't analyse something a "bit more complex" (at least compared to e.g. Rollup).

I would say that we don't worry too much about this for now, as it just feels a waste of time and mental health!
